### ISSUE TITLE
Send/Receive on hold when a FLEx 8 project has FLEx 9 commits in Language Depot history

### ISSUE THREAD

**Issue by rmunn** [sillsdev/LfMerge#107]
At some point in the FLEx 9 development process, the Chorus branch name format went from `7000072` to `7500002.7000072`. The version of Chorus used in lfmerge-7000072 is able to parse that branch format, but the versions of Chorus used in lfmerge-7000070 and earlier use `System.Int32.Parse` to parse branch names, which throws "System.FormatException: Input string was not in the correct format" and puts the project on hold.

We have code in LfMerge to detect that a newer model was pulled and re-run LfMerge with a later model version, but that code is being bypassed by this exception. There are two ways we could fix this issue:

1. Update older Chorus versions used by LfMerge 70000{68,69,70} so that they parse those branch names correctly. Or:

2. LfMerge catches System.FormatException and if it's from System.Int32.Parse, then assume that the model is `7500002.7000072` and just run lfmerge-7000072 to handle that Send/Receive.

The first is the "correct" fix, but the second is the quick fix that we can put into place much more easily. I lean towards doing the second now, and the first if/when we can find the time.

**Comment by rmunn** [comment]
This affects any Language Forge project where the user went from FLEx 8 to FLEx 9. The way the code flows through LfMerge is:

1. src/LfMergeQueueManager/QueueManager.cs starts. For each project in its queue, it calls FwProject.GetModelVersion.
2. FwProject.GetModelVersion looks at the first few lines of the .fwproj file and finds the line `<languageproject version="7000070">`. It extracts the version number from there.
3. QueueManager.cs calls MainClass.StartLfMerge with that model version. (The model version is `null` if no .fwproj file was found, e.g. on an initial clone).
4. MainClass.StartLfMerge finds the `/usr/lib/lfmerge/<modelVersion>/startlfmerge` script (e.g., `/usr/lib/lfmerge/7000070/startlfmerge`) and runs it with command-line args `-p (projectCode) --action (action from queue)`. The "action from queue" depends on which queue the project code was in, e.g. `sync` if it was in the `syncQueue` directory.
5. The `startlfmerge` script sets up environment variables and then runs LfMerge.exe with its command-line arguments.
6. The LfMerge executable parses its command-line arguments and figures out which action it's in (clone, or whatever).
7. Eventually LfMerge gets around to the SynchronizeAction, and calls its DoRun() method.
8. SynchronizeAction.DoRun logs some messages, sets up some counters for entry counts, and calls LfMergeBridge.LfMergeBridge.Execute to do the work. (This is part of Chorus).
9. From LfMerge's point of view, Chorus throws an unhandled System.FormatException and the exception handler in src/LfMerge/Program.cs catches it (inside the RunAction method). The project is then put on hold.

Here's what that exception looks like in practice:

```
System.FormatException: Input string was not in the correct format
  at System.Int32.Parse (System.String s) [0x00012] in /build/mono-sil-3.4.1/mcs/class/corlib/System/Int32.cs:627 
  at LfMergeBridge.LanguageForgeSendReceiveActionHandler.LibTriboroughBridgeChorusPlugin.Infrastructure.IBridgeActionTypeHandler.StartWorking (IProgress progress, System.Collections.Generic.Dictionary`2 options, System.String& somethingForClient) [0x00000] in <filename unknown>:0 
  at LfMergeBridge.LfMergeBridge.Execute (System.String actionType, IProgress progress, System.Collections.Generic.Dictionary`2 options, System.String& somethingForClient) [0x00000] in <filename unknown>:0 
  at LfMerge.Core.Actions.SynchronizeAction.DoRun (ILfProject project) [0x00149] in /build/lfmerge-7000070-1.5.38.98/src/LfMerge.Core/Actions/SynchronizeAction.cs:96 
  at LfMerge.Core.Actions.Action.Run (ILfProject project) [0x00064] in /build/lfmerge-7000070-1.5.38.98/src/LfMerge.Core/Actions/Action.cs:124
```

Normally if lfmerge-7000070 gets a model version of 7000072 from Chorus, then step 8 would have detected the string "pulled a higher model" and then called `ChorusHelper.SetModelVersion(modelVersion)` and then returned. This would have caused the code in the RunAction method to trigger `true` in the `if (ChorusHelper.RemoteDataIsForDifferentModelVersion)` statement, which would then return the newer model version number from RunAction. Then Program.Main would have found a string returned from RunAction, and that string would have caused Program.Main to go call MainClass.StartLfMerge with that different model version.

So what we want to do is catch that exception in step 8, and if it's caught, rerun LfMerge with version 7000072, which is the earliest version that can handle the new branch format. If in the future it turns out that we have a version 7000073, then the rest of the "higher version" code will detect that properly and re-run LfMerge with 7000073.

**Comment by rmunn** [comment]
Here's a diff showing an attempt at solving the issue. I'll test this and if it works I'll turn it into a PR.

```patch
diff --git a/src/LfMerge.Core/Actions/SynchronizeAction.cs b/src/LfMerge.Core/Actions/SynchronizeAction.cs
index 13e1295..f14a611 100644
--- a/src/LfMerge.Core/Actions/SynchronizeAction.cs
+++ b/src/LfMerge.Core/Actions/SynchronizeAction.cs
@@ -93,11 +93,21 @@ protected override void DoRun(ILfProject project)
                                        {"languageDepotRepoUri", chorusHelper.GetSyncUri(project)},
                                        {"commitMessage", commitMessage}
                                };
-                               if (!LfMergeBridge.LfMergeBridge.Execute("Language_Forge_Send_Receive", Progress,
-                                       options, out syncResult))
-                               {
-                                       Logger.Error(syncResult);
-                                       return;
+
+                               try {
+                                       if (!LfMergeBridge.LfMergeBridge.Execute("Language_Forge_Send_Receive", Progress,
+                                               options, out syncResult))
+                                       {
+                                               Logger.Error(syncResult);
+                                               return;
+                                       }
+                               } catch (System.FormatException e) {
+                                       if (e.StackTrace.Contains("System.Int32.Parse")) {
+                                               ChorusHelper.SetModelVersion(MagicStrings.MinimalModelVersionForNewBranchFormat);
+                                               return;
+                                       } else {
+                                               throw;
+                                       }
                                }
 
                                const string cannotCommitCurrentBranch = "Cannot commit to current branch '";
diff --git a/src/LfMerge.Core/MagicStrings.cs b/src/LfMerge.Core/MagicStrings.cs
index e4e9991..140ffdb 100644
--- a/src/LfMerge.Core/MagicStrings.cs
+++ b/src/LfMerge.Core/MagicStrings.cs
@@ -100,6 +100,8 @@ static MagicStrings()
 
                public static string[] UnsupportedModelVersions { get; } = { "7000071" };
 
+               public const int MinimalModelVersionForNewBranchFormat = 7000072;
+
                // Allow to set minimal model version during unit testing
                public static void SetMinimalModelVersion(int minimalModelVersion)
                {
```

**Comment by rmunn** [comment]
I finally managed to build LfMerge locally and test my proposed fix, and I ended up with LfMerge reading the branch name `7500002.7000072` and parsing that as model version 7500002 (the part before the `.`), which is wrong. This ended up trying to run `/usr/lib/lfmerge/7500002/startlfmerge`, which of course doesn't exist. So my fix above, while it appears to work in principle, doesn't work in practice because I'm not actually grabbing the correct part of the model number.

Still, this is progress. All we have to do now is grab the correct model number and we should be okay.
